### PR TITLE
pow scalar exponent / base autodiff, fusion

### LINKED ISSUE #19253: torch.pow() in a script module produces an error
## 🐛 Bug

The backward of torch.pow() in a traced module produces an error when the input is on cuda.

> RuntimeError:
Expected tensor to have CUDA Backend, but got tensor with CPU Backend (while checking arguments for CUDA_tensor_apply4) (checkBackend at ../aten/src/ATen/TensorUtils.cpp:202)

The backward is defined in torch/csrc/jit/symbolic_script.cpp.
I think the reason is that the backend of the first argument of torch.where() is always CPU.

```
def pow_0(self,
          exponent: float):
    def backward(grad_output):
        grad_self = torch.where(torch.tensor(exponent == 0.0), torch.zeros_like(self), grad_output * exponent * torch.pow(self, exponent - 1))
        return grad_self, None
    return torch.pow(self, exponent), backward
```

## To Reproduce

The following model is the one in
https://github.com/pytorch/examples/blob/master/mnist/main.py
but I just inserted torch.pow().

```
import torch
import torch.nn as nn
import torch.nn.functional as F

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 20, 5, 1)
        self.conv2 = nn.Conv2d(20, 50, 5, 1)
        self.fc1 = nn.Linear(4*4*50, 500)
        self.fc2 = nn.Linear(500, 10)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.max_pool2d(x, 2, 2)
        x = F.relu(self.conv2(x))
        x = F.max_pool2d(x, 2, 2)
        x = x.view(64, 4*4*50)
        x = self.fc1(x)
        x = F.relu(x)
        x = self.fc2(x)
        x = torch.pow(x, 1.1)
        x = F.log_softmax(x, dim=1)
        return x

m = Net().cuda()
x = torch.randn(64, 1, 28, 28, requires_grad=True).cuda()
traced_net = torch.jit.trace(m, x)
traced_output = traced_net(x)
tgt = torch.randn(traced_output.size()).cuda()
traced_output.backward(tgt)
```

## Environment
PyTorch version: 1.1.0a0+7e73783
Is debug build: No
CUDA used to build PyTorch: 9.2.88

OS: CentOS Linux release 7.5.1804 (Core)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-28)
CMake version: version 3.12.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration:
GPU 0: Tesla P100-PCIE-16GB
GPU 1: Tesla P100-PCIE-16GB

Nvidia driver version: 396.26
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] numpydoc==0.8.0
[pip] torch==1.1.0a0+7e73783
[pip] torchvision==0.2.2.post3
[conda] blas                      1.0                         mkl
[conda] magma-cuda90              2.5.0                         1    pytorch
[conda] magma-cuda92              2.5.0                         1    pytorch
[conda] mkl                       2019.1                      144
[conda] mkl-include               2019.3                      199
[conda] mkl-service               1.1.2            py37he904b0f_5
[conda] mkl_fft                   1.0.6            py37hd81dba3_0
[conda] mkl_random                1.0.2            py37hd81dba3_0
[conda] torch                     1.1.0a0+7e73783           dev_0    <develop>
[conda] torchvision               0.2.2.post3              pypi_0    pypi

### PR DESCRIPTION
**Pull request by t-vi**
Fixes: #19253

Fixing pow(Tensor, float) is straightforward.
The breakage for pow(float, Tensor) is a bit more subtle to trigger, and fixing needs `torch.log` (`math.log` didn't work) from the newly merged #19115  (Thanks @ngimel for pointing out this has landed.)

### CODE REVIEW

**Comment by t-vi** [comment]
So one of the problems seems to be from Scalar<->float magic in symbolic_script, which seems to break when Scalar is an int. Unfortunately, the only thing I could think of is a gross hack adding torch._float to make a float from int/float IValues passed into a scalar.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (eellison) [inline torch/csrc/jit/register_prim_ops.cpp]: Why does `prim::Float(Scalar a) -> float` not work ?

File: torch/csrc/jit/symbolic_script.cpp
Comment (eellison) [inline torch/csrc/jit/symbolic_script.cpp]: What's preventing float(exponent) instead?

Comment (t-vi) [reply torch/csrc/jit/symbolic_script.cpp]: So what I think is going on: When creating the `pow_0` operator for `Tensor self, Scalar exponent`, symbolic_script replaces Scalar by float. The `float(exponent)` cast gets eliminated because "know" it's float. What then happens is that unpacking Scalar IValues that the JIT thinks must be float but that are, in fact, int fails.

In a way it comes down to
```
    // 2. to make sure the input of any graph node does not contain scalar type
    //    in its argument, all scalar arg should already be passed with float
    //    value since scalar/int aren't differentiable either way.
```
not being the complete picture because (as here) the scalar might not be the thing we want to differentiate for, but a parameter (in the mathematical sense as opposed to the variable) of the function we want to differentiate.

So I'm changing the patch to do the following:
If Scalar -> float conversion happened, I change back the input type of the graph to Scalar, and insert a conversion (`prim::Float`) as the first thing.
It'll be troublesome for use when we get operations that actually rely on the difference between float and int in Scalar ops, but currently we don't as far as I know.

I think this is a clean-up of the Scalar->float conversion as it ensures that the graph inputs actually match the schema.  @ailzhang does that seem reasonable?

File: torch/csrc/jit/register_prim_ops.cpp
Comment (t-vi) [reply torch/csrc/jit/register_prim_ops.cpp]: I needed something to use in symbolic_script, so I'm not sure I have prim ops there. But I've now changed the Scalar<->float conversion to be a bit more prudent about Scalar not always being float.

File: torch/csrc/jit/symbolic_script.cpp
Comment (t-vi) [reply torch/csrc/jit/symbolic_script.cpp]: So it turns out that re-Scalarizing breaks something (the double backward?). :sweat:

File: torch/csrc/jit/register_prim_ops.cpp
Comment (eellison) [reply torch/csrc/jit/register_prim_ops.cpp]: i think the prim ops work - try using a cast - `float(exponent`

File: torch/csrc/jit/symbolic_script.cpp
Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: cc: @wanchaol added the `scalar` to `float` conversion.

File: torch/csrc/jit/register_prim_ops.cpp
Comment (ailzhang) [reply torch/csrc/jit/register_prim_ops.cpp]: @t-vi I feel `float` already works given we register `prim::Float` as `float` here. https://github.com/pytorch/pytorch/blob/ada10ad416f91602d46797b50b3faebcacc6e767/torch/csrc/jit/script/compiler.cpp#L390

File: torch/csrc/jit/symbolic_script.cpp
Comment (eellison) [reply torch/csrc/jit/symbolic_script.cpp]: Also, the scalar to float conversion only happens on the second pass, when allow_conversions is true, if there is an op defined for Scalar than that will be matched and there won't be a conversion.

Comment (ailzhang) [reply torch/csrc/jit/symbolic_script.cpp]: @t-vi  Actually as @eellison pointed out offline, we actually can get rid of the `scalar -> float` conversion entirely. 
For example this works for me. 
```
        def pow_0(self,
                  exponent: number):
``` 
Note that we don't expose `number` in torchscript but we CAN compile it! :D With this we can easily get rid of current `symbolic_variable.cpp` and get rid of `c10::ReplaceAll(schema_str, "Scalar", "float");`. 
Huge thanks to @eellison who pointed this out! 
Let us know if this fixes your problem :D

Comment (t-vi) [reply torch/csrc/jit/symbolic_script.cpp]: Ha. That works (I think)! And I've been poking around much too long. Awesome.
Thanks so much @ailzhang and @eellison !

So my understanding is that with this, one would only have definitions that match the operator schemas. Could we actually check that? For now I left a fallback in but converted those replacements that were hit by a test_jit.py run.

File: torch/csrc/jit/symbolic_script.cpp
Comment (ailzhang) [inline torch/csrc/jit/symbolic_script.cpp]: Yea this `if` should be dropped before merging.

Comment (t-vi) [reply torch/csrc/jit/symbolic_script.cpp]: So I removed it after double checking by means of calling `sig(schema_string)` and seeing whether it throws here:
https://github.com/pytorch/pytorch/blob/3e0b46b6d132c22af84769a6df1ef8d1f785dbe2/torch/csrc/jit/symbolic_script.cpp#L1351-L1353

File: torch/csrc/jit/register_prim_ops.cpp
Comment (t-vi) [reply torch/csrc/jit/register_prim_ops.cpp]: It got eliminated before (because the input claimed it was float). But number fixes it. :)

**Review by ailzhang — APPROVED** [review]
LGTM! Let me know when you are done with changing and want to merge it.

**Comment by t-vi** [comment]
I think its good to merge.

Am 18. April 2019 20:27:33 MESZ schrieb Ailing <notifications@github.com>:

**Pull request closed without merging**
